iceberg/src/main/scala/org/apache/iceberg/spark/functions/bucket.scala

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (c) 2025, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.iceberg.spark.functions
+
+import ai.rapids.cudf.{ColumnVector => CudfColumnVector, DType, Scalar}
+import com.nvidia.spark.rapids.{GpuBinaryExpression, GpuColumnVector, GpuExpression, GpuScalar}
+import com.nvidia.spark.rapids.Arm.withResource
+import com.nvidia.spark.rapids.jni.Hash
+import org.apache.iceberg.spark.functions.GpuBucketExpression.cast
+
+import org.apache.spark.sql.catalyst.expressions.Expression
+import org.apache.spark.sql.types.{ByteType, DataType, DataTypes, DateType, IntegerType, LongType, ShortType, TimestampNTZType, TimestampType}
+
+
+case class GpuBucketExpression(numBuckets: GpuExpression, value: GpuExpression)
+  extends GpuBinaryExpression {
+
+  require(numBuckets.dataType == DataTypes.IntegerType,
+    s"buckets number must be an integer, got ${numBuckets.dataType}")
+
+  require(!value.nullable,
+    s"Bucket function does not support nullable values for type ${value.dataType}")
+
+  require(GpuBucketExpression.isSupportedValueType(value.dataType),
+    s"Bucket function does not support type ${value.dataType} as values")
+
+  override def doColumnar(lhs: GpuColumnVector, rhs: GpuColumnVector): CudfColumnVector = {
+    throw new IllegalStateException("GpuBucketExpression requires first argument to be scalar, " +
+      "second to be column, but both are columns")
+  }
+
+  override def doColumnar(numBuckets: GpuScalar, rhs: GpuColumnVector): CudfColumnVector = {
+    val hash = withResource(cast(rhs.getBase)) { castedValue =>
+      Hash.murmurHash32(0, Array(castedValue))
+    }
+
+    withResource(Scalar.fromInt(Integer.MAX_VALUE)) { intMax =>
+      val nonNegativeHash = withResource(hash) { _ =>
```

Comment:
"hash" will be leaked if the `intMax` fails to be created due to some exception unexpectly. It is rare but it can happen. So
```
withResource(hash) { _ =>
  withResource(Scalar.fromInt(Integer.MAX_VALUE)) { intMax =>
    hash.bitAnd(intMax)
  }
}

```
should be better.